idseq_dag/steps/reclassify_reads.py

```diff
@@ -1,22 +1,248 @@
+                error_flags["error"] = 1
+        finally:
+            semaphore.release()
+
+
+    @staticmethod
+    def assemble_all(genus_fasta_files):
+        genus_assembled = {}
+        for genus_taxid, fasta_file in genus_fasta_files:
+            genus_dir = os.path.dirname(fasta_file)
+            assembled_dir = os.path.join(genus_dir, 'spades')
+            output = (None, None, None) # (contigs.fasta path, scaffolds.fasta path, readid -> contig mapping)
+            command.execute(f"mkdir -p {assembled_dir}")
+
+            try:
+                command.execute(f"spades.py -s {fasta_file} -o {assembled_dir} -m 60 -t 32 --only-assembler")
+                assembled_contig = os.path.join(assembled_dir, 'contigs.fasta')
+                assembled_scaffold = os.path.join(assembled_dir, 'scaffolds.fasta')
+
+                for i, assembled in enumerate([assembled_contig, assembled_scaffold]):
+                    if os.path.exists(assembled) and os.path.getsize(assembled) > MIN_CONTIG_SIZE:
+                        # move the file up to the main dir
+                        final_path = assembled.replace('spades/', '')
+                        command.execute(f"mv {assembled} {final_path}")
+                        output[i] = final_path
+                # build the bowtie index based on the contigs
+                if output[0]:
+                    output[2] = self.generate_read_to_contig_mapping(output[0], fasta_file)
+            except:
+                pass
+            genus_assembled[genus_taxid] = output
+            command.execute(f"rm -rf {assembled_dir}")
+
+        return genus_assembled
+
+    @staticmethod
+    def generate_read_to_contig_mapping(assembled_contig, fasta_file):
+        genus_dir = os.path.dirname(fasta_file)
+        # build bowtie index based on assembled_contig
+        bowtie_index_path = os.path.join(genus_dir, 'bowtie-contig')
+        bowtie_sam = os.path.join(genus_dir, "read-output.sam")
+        command.execute(f"bowtie2-build {assembled_contig} {bowtie_index_path}")
+        command.execute(f"bowtie2 -x {bowtie_index_path} -f -U {fasta_file} --very-sensitive -p 32 > {bowtie_sam}")
+        read2contig = {}
+        with open(bowtie_sam, "r", encoding='utf-8') as samf:
+            for line in samf:
+                if line[0] == '@':
+                    continue
+                fields = line.split("\t")
+                read = fields[0]
+                contig = fields[2]
+                if contig != '*':
+                    read2contig[read] = contig
+        return read2contig
+
+
+    @staticmethod
+    def group_reads_by_genus(input_fasta, read_dict, selected_genera, output_basedir):
+        ''' returns genus_id => fasta file which includes sequences from the genus from alignment '''
+        genus_fasta_data = defaultdict("")
+        genera_list = set(selected_genera.keys())
+        with open(input_fasta, 'r', encoding='utf-8') as input_fasta_f:
```

Comment:
I think you should use the fasta iterator Boris wrote: https://github.com/chanzuckerberg/idseq-dag/blob/4214c88ec6c9114ffce4f96c896f27fd043065b4/idseq_dag/util/fasta.py#L9